release: Add v1.0.0 changelog (#184)
elastic/cloud-sdk-go: notes/v1.0.0.md

# Changelog

## Breaking changes

### Improved api default values

Changes the constructor behavior of `NewAPI` and the `Config`
validation now only accepts an `auth.APIKey` or `*auth.APIKey` as the

> only valid AuthWriter's for the API when the host is set to the ESS host

Reviewer: ```suggestion
only valid AuthWriters for the API when the host is set to the ESS host
```

> (`https://api.elastic-cloud.com`). Also adding a constant: `ESSEndpoint`.

Reviewer: ```suggestion
(`https://api.elastic-cloud.com`). Also adds a constant: `ESSEndpoint`.
```

Adds a new function to the `deploymentapi` pkg: `deploymentapi.RequestID`
which will always return a request ID to be
used in `deploymentapi.CreateParams`. It will either return an auto-
generated string of 64 characters or the string argument which is passed
to the function.

Adds a new package called `deploymentsize` and a new function:

> `Parse` which will parse a string in the format of `<size>g`, valid

Reviewer: ```suggestion
`ParseGB` which will parse a string in the format of `<size>g` into its value in megabytes as an int32. Valid
```

> examples of parseable sizes are in `0.5g` increments with the minimum

Reviewer: ```suggestion
examples of parseable sizes are in `0.5g` increments, with a minimum
```

accepted size of `0.5g`.

### Region is no longer a valid `api.Config` parameter

The `Region` field on `api.Config` has been removed from the structure.
Instead of using a global Region parameter, all APIs which require a region
to be specified as a parameter, will expose a `Region` parameter to be used.

## Notable changes

### Changes to existing APIs

#### Added PayloadOverrides to deploymentapi.Update()

Modified the Update function parameters to allow the consumer to specify
a PayloadOverrides which can be used to override nested values in the resources.
Currently supports overrides for `Version`, `Name` and `Region`.

> #### Support redacted Authorization headers

Reviewer: ```suggestion
#### Support redacted authorization headers
```

Added a new configuration parameter `RedactAuth` to `api.VerboseSettings`
when set, will replace the contents of the `Authorization` header with
`[REDACTED]`.

The default behavior still preserves the original contents of the header.

#### Support for Enterprise Search on all APIs

> Adds support for the `enterprise_search` workload type on all APIs, namely:

Reviewer: ```suggestion
Adds support for the `enterprise_search` resource kind on all APIs, namely:
```

* [`pkg/api/platformapi/constructorapi`](https://pkg.go.dev/github.com/elastic/cloud-sdk-go/pkg/api/platformapi/constructorapi)
* [`pkg/api/stackapi`](https://pkg.go.dev/github.com/elastic/cloud-sdk-go/pkg/api/stackapi)
* [`pkg/api/userapi`](https://pkg.go.dev/github.com/elastic/cloud-sdk-go/pkg/api/userapi)
* [`pkg/api/userapi/authapi`](https://pkg.go.dev/github.com/elastic/cloud-sdk-go/pkg/api/userapi/authapi)
* [`pkg/api/userapi/authapi/adminapi`](https://pkg.go.dev/github.com/elastic/cloud-sdk-go/pkg/api/userapi/authapi/adminapi)
* [`pkg/api/deploymentapi`](https://pkg.go.dev/github.com/elastic/cloud-sdk-go/pkg/api/deploymentapi)
* [`pkg/api/deploymentapi/eskeystoreapi`](https://pkg.go.dev/github.com/elastic/cloud-sdk-go/pkg/api/deploymentapi/eskeystoreapi)
* [`pkg/api/deploymentapi/noteapi`](https://pkg.go.dev/github.com/elastic/cloud-sdk-go/pkg/api/deploymentapi/noteapi)
* [`pkg/api/deploymentapi/depresourceapi`](https://pkg.go.dev/github.com/elastic/cloud-sdk-go/pkg/api/deploymentapi/depresourceapi)
* [`pkg/api/deploymentapi/deptemplateapi`](https://pkg.go.dev/github.com/elastic/cloud-sdk-go/pkg/api/deploymentapi/deptemplateapi)

> #### New deptemplateapi package

Reviewer: Do we need to mention this package? It's already on the list above 🤔

Added all the APIs for the new deployment templates API which present the

> deployment templates in the "deployment" format definitions. Which are
> closer to the deployment create and update request payloads.

Reviewer: ```suggestion
deployment templates in the "deployment" format definitions. These APIs
use consume and produce structures which are closer to the deployment 
create and update request payloads.
```

Reviewer: I kind of want to remove that last sentence all-together as people reading this won't know about the old deployment-template endpoints so it might sound confusing to say something is "closer to...". As in, closer than what?

Author: I think that makes sense, I'll remove that line

> #### New eskeystoreapi package

Reviewer: Do we need to mention this package? It's already on the list above 🤔

Adds a new package `eskeystoreapi` which implements the new keystore API
under the deployments API. With ref_id auto-discovery to both `Get` and
`Update` functions, and a new helper function `PopulateRefID()` which
makes adding resource auto-discovery less logic heavy by populating a
`*string`, needing to handle less on the implementation side.

#### Added a api.WithRegion() function, allowing to set a region per API call

Added a `context.Context` constructor (`api.WithRegion`) which creates a
new context with a saved value equal to the passed region. This is meant
to be used in any calls to the auto-generated API which are region-bound
(any `/platform` APIs).

When using `*api.CloudClientRuntime`, the path interpolation will take
place without the need to explicitly handle it.

This change allows the `*api.API` object to be used accross multiple
regions rather than requiring a single instance of that structure per
region.

### Bug fixes

#### API now handles certain API call string fields as "raw JSON"

Adds certain handling of API operation string fields to be handled as raw
JSON fields so that double quotes in said fields aren't escaped by the

> runtime, the excluded operations are:

Reviewer: ```suggestion
runtime. The excluded operations are:
```

### Testing

#### Added support for request assertions on unit tests.

Adds the ability to assert certain `*http.Request` fields which
are sent to the http.RoundTripper on any of the request / reply flows
which are tested and asserted in our unit tests. Since it's an optional
field, it doesn't enforce that field to be set in the `mock.Response`,
but rather make it optional for those tests where it makes sense to do so.

#### Added `apivalidator` testing layer

Added a new `apivalidator` executable:

- A CLI running a series of requests based off a specified apidocs file
(which can be either a local file or downloaded from it's location), against
a prism API validation proxy, and then returns errors based on discrepancies.

- A new Makefile with targets to start a prism validation proxy for the public
API, another to run the validation tests  against the proxy, and another to run
both without having to start a new terminal and killing the server when it finishes.

Additionally, these tests are now run on every `make swagger` to verify endpoint
availability when new `client` and `models` are generated from a swagger spec.

### Detailed changes
<!-- AUTOGENERATED COMMIT LIST -->

> 

Reviewer: can we add the latest commit with the bug fix for `deploymentsize.ParseGB()`?
